src/dotnet.ts

```diff
@@ -0,0 +1,135 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as vscode from 'vscode';
+import * as which from 'which';
+import * as cp from 'child_process';
+import * as semver from 'semver';
+import * as path from 'path';
+import * as Constants from './constants';
+import { CommandObserver } from './CommandObserver';
+
+export type DotNetInfo = {path: string, version: string};
+export type DotNetCommandResult = {code: number, msg: string};
+
+let dotnetInfo: DotNetInfo | undefined = undefined;
+
+function promptToInstallDotNetCoreSDK(msg: string): void {
+    let installItem = 'Install .NET Core SDK...';
+    vscode.window
+        .showErrorMessage(msg, installItem)
+        .then(
+            (item) => {
+                if (item === installItem) {
+                    vscode.env.openExternal(vscode.Uri.parse('https://dotnet.microsoft.com/download'));
+                }
+            }
+        );
+}
+
+/**
+ * Returns the path to the .NET Core SDK, or prompts the user to install
+ * if the SDK is not found.
+ */
+export function findDotNetSdk(): Promise<DotNetInfo> {
+    return new Promise((resolve, reject) => {
+        if (dotnetInfo === undefined) {
+            try {
+                let dotnetPath = which.sync('dotnet');
+                cp.exec(
+                    'dotnet --version',
+                    (error, stdout, stderr) => {
+                        if (error) {
+                            reject(error);
+                        } else if (stderr && stderr.length > 0) {
+                            reject(new Error(stderr));
+                        } else {
+                            dotnetInfo = {path: dotnetPath, version: stdout.trim()};
+                            resolve(dotnetInfo);
+                        }
+                    }
+                );
+            } catch (ex) {
+                promptToInstallDotNetCoreSDK('The .NET Core SDK was not found.');
+                reject(ex);
+            }
+        } else {
+            resolve(dotnetInfo);
+        }
+    });
+}
+
+export function findProjectTemplate(dotNetSdk: DotNetInfo): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+        let exists: boolean = true;
+        let cmd = 'dotnet';
+        let args = ['new', 'pgproject', '--dry-run'];
+        let dotnet = cp.spawn(cmd, args, { cwd: path.dirname(dotNetSdk.path), env: process.env });
```

Comment:
Hm.. is there a way to create the class (without starting the process), then hook to the events and streams and then start it? I'm wondering what is going to happen if it exits before you hook the ```close``` event? I think there is a change we will never resolve the promise and it will hang forever...

Reply:
Looking at the ```child_process``` [source code](https://searchcode.com/codesearch/view/89235565/), I'm not sure I understand how it is supposed to work. They do create new ```ChildProcess``` instance and spawn it when you call ```spawn```, then you hook to events, which are hooked through ```EventEmitter``` base class... Unless it buffers them - there is indeed a chance to miss the exit/close. I doubt it does any buffering, because that will be a potential OOM.
It seems like you can just create ```ChildProcess``` instance yourself, hook the events and then call ```spawn``` on it, but they also do ```normalizeSpawnArguments```, which you will need. It doesn't do much and we probably don't need a half of it.
It could be that the assumption is on the nodejs being a single-threaded - they raise events from async handlers, which should not fire before your linear code here finishes execution and yields the thread. You do ```cp.spawn```, that internally does ```this._handle.onexit = function(exitCode, signalCode)...```, which emit the ```exit```/```close``` event, but it can't call ```onexit``` handler until the code in this block finishes execution. I suppose it works today, but I don't know how stable is this assumption...